add a compact epoch ID called `epoclet` (#71)
ietf-rats/draft-birkholz-rats-epoch-marker: draft-ietf-rats-epoch-markers.md

Systems that use Epoch Markers SHOULD follow the guidance in {{sec-state-seq-mgmt}} in establishing an Epoch Marker acceptance policy for receivers.
To prove freshness, receivers SHOULD track the highest accepted counter and ensure it fulfills the acceptance policy.

> ### Stateless Nonce {#sec-stateless-nonce}

Reviewer: By stateless do you mean the nonce itself is stateless, the bell is stateless, or both?

Author: I couldn’t come up with anything better than this rubbish name, sorry :-D
Maybe “tiny epoch-id”?
Anyway, feel free to suggest a better name!

Reviewer: I was mostly curious if the implication that the nonce is stateless too is correct (?!). Otherwise "Stateless Bell Nonce" might be more accurate, but it's a mouthful ☹️ 

Happy to go ahead as-is, btw

Author: To divert attention away from a detail that I suspect will be a never-ending source of bikeshedding, I’ve come up with “epoclet”, a completely absurd and meaningless name.

Reviewer: > from a detail that I suspect will be a never-ending source of bikeshedding

Sounds like you have some experience with that!

> epoclet

Okay, new word just dropped, let's go! 😎

In a highly available service (e.g., a cloud attestation Verifier), maintaining per-session nonce state can cause scalability issues.
One alternative is to use time-synchronized servers that share a symmetric key and produce and consume nonces based on coarse-grained clock ticks signed using the shared secret.
This means that a nonce minted by one server can be processed by any other server, avoiding the need for session "stickiness".

> A `stateless-nonce` supports the above use case by encoding a Posix time (i.e., the epoch identifier) alongside a minimal set of metadata.

Reviewer 2: ```suggestion
A `stateless-nonce` supports the above use case by encoding a POSIX time (i.e., the epoch identifier) alongside a minimal set of metadata.
```

Systems that use Epoch Markers SHOULD follow the guidance in {{sec-state-seq-mgmt}} in establishing an Epoch Marker acceptance policy for receivers.
To prove freshness, receivers SHOULD track the highest accepted counter and ensure it fulfills the acceptance policy.

### Epoclet {#sec-epoclet}

In a highly available service (e.g., a cloud attestation Verifier), maintaining per-session nonce state can cause scalability issues.

> One alternative is to use time-synchronized servers that share a symmetric key and produce and consume nonces based on coarse-grained clock ticks signed using the shared secret.

Reviewer 2: maybe?
```suggestion
One alternative is to use time-synchronized servers that share a symmetric key and produce and consume nonces based on epoch ticks signed using the shared secret.
```

Author: The thing about “coarse-grained” is that “epoclet”s have 1s resolution, at best.

This means that a nonce minted by one server can be processed by any other server, avoiding the need for session "stickiness".

An `epoclet` is an Epoch ID variant that supports the above use case by encoding a POSIX time (i.e., the epoch identifier) alongside a minimal set of metadata.
This is all authenticated with a symmetric key in a self-contained and compact token that fits within 64 bytes.
This makes it easy to use with common Evidence retrieval ABIs, which tend to limit the size of the challenge parameter to 64 bytes.

~~~~ cddl
{::include cddl/epoclet.cddl}
~~~~

The following describes the `epoclet` type.

{: vspace="0"}
KeyID:
: A one-byte opaque identifier that is shared across the server pool.
It is used to identify the key used to compute the AuthTag.
Its interpretation is deployment-specific.

Timestamp:
: The time associated with the current epoch encoded as the CBOR tag for POSIX time.
It MUST use the int format.
The CBOR tag MUST be removed.
Note that this encoding restricts the granularity of this Epoch ID to one second.

Pad:
: Zero or more (up to 20) bytes of padding.

> This field is used to adjust the overall size of the nonce, ranging from a minimum of 44 bytes (0 bytes of padding) to a maximum of 64 bytes (20 bytes of padding).

Author: ```suggestion
This field is used to adjust the overall size of the epoclet, ranging from a minimum of 44 bytes (0 bytes of padding) to a maximum of 64 bytes (20 bytes of padding).
```

> This field is used to adjust the overall size of the epoclet, ranging from a minimum of 44 bytes (0 bytes of padding) to a maximum of 64 bytes (20 bytes of padding).

Author: ```suggestion
This field is used to adjust the overall size of the `epoclet`, ranging from a minimum of 44 bytes (0 bytes of padding) to a maximum of 64 bytes (20 bytes of padding).
```

The padding bytes can assume any value.

AuthTag:

> : 32 bytes string which is the result of applying HMAC {{!RFC2104}} w/ SHA-256 over the CBOR serialization of the TimeToken array.

Reviewer: ```suggestion
: 32 bytes string which is the result of applying HMAC {{!RFC2104}} with SHA-256 over the CBOR serialization of the TimeToken array.
```

Am I missing some pre-processing magic?

Author: 🤔 not sure I get your point

Author: I had something like this in mind:

```go
timeToken = &TimeToken{
   KeyID: keyID,
   Timestamp: time.Now(),
   Pad: pad,
}

v, _ := timeToken.MarshalCBOR()

epoclet = &Epoclet{
  TimeToken: timeToken,
  AuthTag: HMAC(v, k)
}

return epoclet.MarshalCBOR()
```
